File: blurb/app/src/main/java/com/mowdowndevelopments/blurb/ui/login/LoginViewModel.java
package com.mowdowndevelopments.blurb.ui.login;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.mowdowndevelopments.blurb.R;
import com.mowdowndevelopments.blurb.network.LoadingStatus;
import com.mowdowndevelopments.blurb.network.ResponseModels.AuthResponse;
import com.mowdowndevelopments.blurb.network.Singletons;

import org.jetbrains.annotations.NotNull;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import timber.log.Timber;

public class LoginViewModel extends AndroidViewModel {

    private MutableLiveData<LoadingStatus> loginStatus;
    private MutableLiveData<String> errorToast;

    private final Callback<AuthResponse> loginCallback= new Callback<AuthResponse>() {
        @Override
        public void onResponse(@NotNull Call<AuthResponse> call, @NotNull Response<AuthResponse> response) {
            if (response.isSuccessful()) {
                if (response.body() != null && response.body().isAuthenticated()) {
                    loginStatus.postValue(LoadingStatus.DONE);
                } else {
                    loginStatus.postValue(LoadingStatus.ERROR);
                    errorToast.postValue(getApplication().getString(R.string.bad_credential_error));
                    Timber.w("loginCallback.onResponse: User entered incorrect credentials.");
                }
            } else {
                loginStatus.postValue(LoadingStatus.ERROR);
                //TODO Add messages to explain HTTP Error codes in plain language
                String errorMsg = getApplication().getString(R.string.http_error, response.code());
                errorToast.postValue(errorMsg);
                Timber.e("loginCallback.onResponse: %s", errorMsg);
            }
        }

        @Override
        public void onFailure(@NotNull Call<AuthResponse> call, @NotNull Throwable t) {
            loginStatus.postValue(LoadingStatus.ERROR);
            Timber.e(t, "loginCallback.onFailure: %s", t.getMessage());
            errorToast.postValue(t.getLocalizedMessage());
        }
    };
    private final Callback<AuthResponse> registrationCallback= new Callback<AuthResponse>() {
        @Override
        public void onResponse(@NotNull Call<AuthResponse> call, @NotNull Response<AuthResponse> response) {
            if (response.isSuccessful()) {
                if (response.body() != null && response.body().isAuthenticated()) {
                    loginStatus.postValue(LoadingStatus.DONE);
                } else {
                    loginStatus.postValue(LoadingStatus.ERROR);
                    errorToast.postValue(getApplication().getString(R.string.bad_registration_error));
                    Timber.w("registrationCallback.onResponse: User entered incorrect credentials.");
                }
            } else {
                loginStatus.postValue(LoadingStatus.ERROR);
                //TODO Add messages to explain HTTP Error codes in plain language
                String errorMsg = getApplication().getString(R.string.http_error, response.code());
                errorToast.postValue(errorMsg);
                Timber.e("loginCallback.onResponse: %s", errorMsg);
            }
        }

        @Override
        public void onFailure(@NotNull Call<AuthResponse> call, @NotNull Throwable t) {
            loginStatus.postValue(LoadingStatus.ERROR);
            Timber.e(t, "loginCallback.onFailure: %s", t.getMessage());
            errorToast.postValue(t.getLocalizedMessage());
        }
    };

    public LiveData<LoadingStatus> getLoginStatus(){ return loginStatus; }

    public LiveData<String> getErrorToast() {
        return errorToast;
    }

    public LoginViewModel(@NonNull Application application) {
        super(application);
        loginStatus = new MutableLiveData<>(LoadingStatus.WAITING);
        errorToast = new MutableLiveData<>();
    }

    public void login(String username, String password){
        if (loginStatus.getValue() == LoadingStatus.LOADING) return;
        loginStatus.postValue(LoadingStatus.LOADING);
        Singletons.getNewsBlurAPI(getApplication()).login(username, password).enqueue(loginCallback);
    }

    public void login(String username){
        if (loginStatus.getValue() == LoadingStatus.LOADING) return;
        loginStatus.postValue(LoadingStatus.LOADING);
        Singletons.getNewsBlurAPI(getApplication()).login(username).enqueue(loginCallback);
    }

    public void registerNewAccount(String username, String password, String emailAddress){
        if (loginStatus.getValue() == LoadingStatus.LOADING) return;
        loginStatus.postValue(LoadingStatus.LOADING);
        Singletons.getNewsBlurAPI(getApplication()).signup(username, password, emailAddress).enqueue(registrationCallback);
    }

    public void registerNewAccount(String username, String emailAddress){
        if (loginStatus.getValue() == LoadingStatus.LOADING) return;
        loginStatus.postValue(LoadingStatus.LOADING);
        Singletons.getNewsBlurAPI(getApplication()).signup(username, emailAddress).enqueue(registrationCallback);
    }
}

File: blurb/app/src/main/java/com/mowdowndevelopments/blurb/ui/login/LoginViewModel.kt
package com.mowdowndevelopments.blurb.ui.login

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import com.google.firebase.crashlytics.FirebaseCrashlytics
import com.mowdowndevelopments.blurb.BlurbApp
import com.mowdowndevelopments.blurb.R
import com.mowdowndevelopments.blurb.network.LoadingStatus
import com.mowdowndevelopments.blurb.network.Singletons.getNewsBlurAPI
import com.mowdowndevelopments.blurb.network.responseModels.AuthResponse
import kotlinx.coroutines.launch
import retrofit2.Response
import timber.log.Timber

class LoginViewModel(application: Application) : AndroidViewModel(application) {

    private val _loginStatus: MutableLiveData<LoadingStatus> = MutableLiveData(LoadingStatus.WAITING)
    val loginStatus: LiveData<LoadingStatus>
        get() = _loginStatus
    private val _errorToast: MutableLiveData<String> = MutableLiveData()
    val errorToast: LiveData<String>
        get() = _errorToast

    fun login(username: String, password: String) {
        if (_loginStatus.value === LoadingStatus.LOADING) return
        viewModelScope.launch {
            _loginStatus.postValue(LoadingStatus.LOADING)
            try {
                val response = getNewsBlurAPI(getApplication()).login(username, password)
                checkLoginAuthResponse(response)
            } catch (e: Exception) {
                authFailure(e)
            }
        }
    }

    fun login(username: String) {
        if (_loginStatus.value === LoadingStatus.LOADING) return
        viewModelScope.launch {
            _loginStatus.postValue(LoadingStatus.LOADING)
            try {
                val response = getNewsBlurAPI(getApplication()).login(username)
                checkLoginAuthResponse(response)
            } catch (e: Exception) {
                authFailure(e)
            }
        }
    }

    private fun checkLoginAuthResponse(response: Response<AuthResponse>) {
        val authenticated = response.body()?.isAuthenticated
        when {
            response.isSuccessful && authenticated!! -> _loginStatus.postValue(LoadingStatus.DONE)
            response.isSuccessful && !authenticated!! -> {
                _loginStatus.postValue(LoadingStatus.ERROR)
                _errorToast.postValue(getApplication<BlurbApp>().getString(R.string.bad_credential_error))
                Timber.w("loginCallback.onResponse: User entered incorrect credentials.")
            }
            !response.isSuccessful -> {
                _loginStatus.postValue(LoadingStatus.ERROR)
                //TODO Add messages to explain HTTP Error codes in plain language
                val errorMsg = getApplication<BlurbApp>().getString(R.string.http_error, response.code())
                _errorToast.postValue(errorMsg)
                Timber.e("loginCallback.onResponse: $errorMsg")
            }
            else -> {
                _loginStatus.postValue(LoadingStatus.ERROR)
                val errorMsg = getApplication<BlurbApp>().getString(R.string.msg_unknown_error)
                _errorToast.postValue(errorMsg)
                Timber.e("loginCallback.onResponse: $errorMsg")
            }
        }
    }

    fun registerNewAccount(username: String, password: String, emailAddress: String) {
        if (_loginStatus.value === LoadingStatus.LOADING) return
        viewModelScope.launch {
            _loginStatus.postValue(LoadingStatus.LOADING)
            try {
                val response = getNewsBlurAPI(getApplication()).signup(username, password, emailAddress)
                checkRegisterAuthResponse(response)
            } catch (e: Exception) {
                authFailure(e)
            }
        }
    }

    fun registerNewAccount(username: String, emailAddress: String) {
        if (_loginStatus.value === LoadingStatus.LOADING) return
        viewModelScope.launch {
            _loginStatus.postValue(LoadingStatus.LOADING)
            try {
                val response = getNewsBlurAPI(getApplication()).signup(username, emailAddress)
                checkRegisterAuthResponse(response)
            } catch (e: Exception) {
                authFailure(e)
            }
        }
    }

    private fun checkRegisterAuthResponse(response: Response<AuthResponse>) {
        val authenticated = response.body()?.isAuthenticated
        when {
            response.isSuccessful && authenticated == true -> _loginStatus.postValue(LoadingStatus.DONE)
            response.isSuccessful && authenticated == false -> {
                _loginStatus.postValue(LoadingStatus.ERROR)
                _errorToast.postValue(getApplication<Application>().getString(R.string.bad_registration_error))
                Timber.w("registrationCallback.onResponse: User entered incorrect credentials.")
            }
            !response.isSuccessful -> {
                _loginStatus.postValue(LoadingStatus.ERROR)
                //TODO Add messages to explain HTTP Error codes in plain language
                val errorMsg = getApplication<BlurbApp>().getString(R.string.http_error, response.code())
                _errorToast.postValue(errorMsg)
                Timber.e("registrationCallback.onResponse: $errorMsg")
            }
            else -> {
                _loginStatus.postValue(LoadingStatus.ERROR)
                val errorMsg = getApplication<Application>().getString(R.string.msg_unknown_error)
                _errorToast.postValue(errorMsg)
                Timber.e("registrationCallback.onResponse: $errorMsg")
            }
        }
    }

    private fun authFailure(e: Exception) {
        _loginStatus.postValue(LoadingStatus.ERROR)
        Timber.e(e, "onFailure: %s", e.message)
        _errorToast.postValue(e.localizedMessage)
        FirebaseCrashlytics.getInstance().log(String.format("onFailure: %s", e.message))
        FirebaseCrashlytics.getInstance().recordException(e)
    }

}